nexus/db-queries/src/db/datastore/oximeter_reads_policy.rs

```diff
@@ -0,0 +1,171 @@
+    ) -> Result<Option<OximeterReadsPolicy>, Error> {
+        opctx.authorize(authz::Action::Read, &authz::BLUEPRINT_CONFIG).await?;
+        let conn = self.pool_connection_authorized(opctx).await?;
+
+        use db::schema::oximeter_reads_policy::dsl;
+
+        let latest_policy = dsl::oximeter_reads_policy
+            .order_by(dsl::version.desc())
+            .first_async::<DbOximeterReadsPolicy>(&*conn)
+            .await
+            .optional()
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(latest_policy.map(Into::into))
+    }
+
+    /// Insert the current version of the policy in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently
+    /// in the `oximeter_reads_policy` table. If there are no versions currently
+    /// in the table, then the current policy must be at version 1.
+    pub async fn oximeter_reads_policy_insert_latest_version(
+        &self,
+        opctx: &OpContext,
+        policy: &OximeterReadsPolicy,
+    ) -> Result<(), Error> {
+        if policy.version < 1 {
+            return Err(Error::invalid_request(
+                "policy version must be greater than 0",
+            ));
+        }
+        opctx
+            .authorize(authz::Action::Modify, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let num_inserted = if policy.version == 1 {
+            self.oximeter_reads_policy_insert_first_policy(opctx, &policy)
+                .await?
+        } else {
+            self.oximeter_reads_policy_insert_next_policy(opctx, &policy)
+                .await?
+        };
+
+        match num_inserted {
+            0 => Err(Error::invalid_request(format!(
+                "policy version {} is not the most recent",
+                policy.version
+            ))),
+            1 => Ok(()),
+            // This is impossible because we are explicitly inserting only one
+            // row with a unique primary key.
+            _ => unreachable!("query inserted more than one row"),
+        }
+    }
+
+    /// Insert the next version of the policy in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently
+    /// in the `oximeter_reads_policy` table.
+    ///
+    /// Panics if `policy.version <= 1`;
+    async fn oximeter_reads_policy_insert_next_policy(
+        &self,
+        opctx: &OpContext,
+        policy: &OximeterReadsPolicy,
+    ) -> Result<usize, Error> {
+        assert!(policy.version > 1);
+        let prev_version = policy.version - 1;
+
+        sql_query(
+                r"INSERT INTO oximeter_reads_policy
+                     (version, oximeter_reads_mode, time_created)
+                     SELECT $1, $2, $3
+                      FROM oximeter_reads_policy WHERE version = $4 AND version IN
+                       (SELECT version FROM oximeter_reads_policy
+                        ORDER BY version DESC LIMIT 1)",
+            )
```

Comment:
Totally up to you, but if you want to stick to the diesel typed universe, @jgallagher just showed me how to wrangle basically this same query in [#7518](https://github.com/oxidecomputer/omicron/pull/7518#discussion_r1951210349).